### PR TITLE
Add unit test coverage for Installer

### PR DESCRIPTION
**Pull request by yariksheptykin**
This PR is an attempt to add some unit test converage for Installer

Complementes https://github.com/pimcore/pimcore/pull/6897

### CODE REVIEW

**Comment by yariksheptykin** [comment]
@dvesh3 @fashxp @jdreesen after playing around with existing unit tests and the implementation of the installer it became clear to me that:

1. simply adding tests for #6897 is not possible without testing the whole installer because the functionality is deeply incapsulated,
2. Testing installer as a whole with current approach to unit tests is going to be very difficult because the istaller will alter the complete testing environment. Installer (end-to-end) test thus must run in isolation

To address 1) a suggest pulling the affected method (`createConfigFiles`) out of intaller into a stand alone class or into `ConfigWriter`. This will make testing of this single unit of code possible. What do you think? To work around 1) i temporarily made the method public to demonstrate how a unit test for this can look like.

Further UnitTests for Installer would require a slight refactoring of its usage of static calls and dependency management. This is necessary to provide isolated testing environment for it.

**Comment by yariksheptykin** [comment]
FYI, i was using this setup to exexute tests:
- test environment set up with docker compose:
```yaml
 version: '2.4'

services:
    redis:
        image: redis:alpine

    db:
        image: mariadb:10.4
        working_dir: /application
        command: [mysqld, --character-set-server=utf8mb4, --collation-server=utf8mb4_unicode_ci, --innodb-file-format=Barracuda, --innodb-large-prefix=1, --innodb-file-per-table=1]
        environment:
            - MYSQL_ROOT_PASSWORD=pimcore
            - MYSQL_DATABASE=pimcore_test
            - MYSQL_USER=pimcore
            - MYSQL_PASSWORD=pimcore

    php:
        image: pimcore/pimcore:PHP7.4-apache-debug
        volumes:
            - .:/var/www/html:cached
        hostname: pimcore-test.dev
        depends_on:
          - db
          - redis
        environment:
            - COMPOSER_MEMORY_LIMIT=-1
            - TASK_TESTS=1
            - TASK_STAN=0
            - TASK_DOCS=0
            - PHPSTAN_LEVEL=3
            - PHPSTAN_BASELINE=1
            - PHPSTAN_BASELINE_GENERATE=0
            - PIMCORE_ENVIRONMENT=test
            - PIMCORE_TEST=1
            - PIMCORE_TEST_URL=http://pimcore-test.dev
            - PIMCORE_TEST_DB_DSN="mysql://root:pimcore@db/pimcore_test"
            - PIMCORE_TEST_CACHE_REDIS_DATABASE=1
            - SYMFONY_VERSION=^4.0
            - PIMCORE_PROJECT_ROOT=/var/www/pimcore
            - PIMCORE_PHP_ERROR_REPORTING=32767 # E_ALL
```
- `docker-compose exec php sh .travis/setup-tests.sh`
- `docker-compose exec php php vendor/bin/codecept run -c . Unit`

**Comment by yariksheptykin** [comment]
@dvesh3 , @fashxp , @jdreesen , what do you think, is this MR moving in a right direction? Should I carry on or drop it? I defninitely don't want to keep blocking https://github.com/pimcore/pimcore/pull/6897

**Comment by dvesh3** [comment]
@yariksheptykin I have marked this "pending" for internal discussion. will update you soon 😉

**Comment by dvesh3** [comment]
@yariksheptykin 
1. we can go ahead with suggested refactoring to `ConfigWriter` and also more refactoring required for isolated testing environment.
2. please create a dedicated job to run installer tests in isolation like for docs and stan,https://github.com/pimcore/pimcore/blob/e9a1b63c812882dbddbb5bc6be122c17b8b6f26f/.travis.yml#L220

**Comment by dvesh3** [comment]
@yariksheptykin We merged the PR https://github.com/pimcore/pimcore/pull/6897 to master so I changed the target base to `master` now. Please rebase and cherry pick your commits. Now, you have more time to work on this PR 😉

**Comment by yariksheptykin** [comment]
@dvesh3 I believe I this MR is ready for review. I added test for the install command as a whole. It runs the complete installation as `bin/pimcore-install` would do. I only validated creation of system files in this test for now.
I also extracted the method for creating system configuration files into a stand alone class and added a simple unit test for it.

**Comment by dvesh3** [comment]
@yariksheptykin Amazing work 🙌 specially with the detailed comments. I'll review and test it soon. 

just one concern, any reason for having only Symfony 3.4 tests?

**Comment by yariksheptykin** [comment]
> @yariksheptykin Amazing work raised_hands specially with the detailed comments. I'll review and test it soon.

Thanks. It might look good, but it's quite hacky (see copying migrations) because I not familliar with installer code to set up a better fixture. I tried to complement the hacky-implementation with comments to explain what is going on.

> just one concern, any reason for having only Symfony 3.4 tests?

@dvesh3  no, that was arbitrary. I copied the top most job in the .travis-file as a template for installer tests. We can freely change 
this or add more environments.

File: tests/Installer/Command/InstallCommandTest.php [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+     * @See self::disposeFile()
+     */
+    private function restoreFile(string $filePath): void
+    {
+        $this->moveIfExits($this->createBackupFileName($filePath), $filePath);
+    }
+
+    /**
+     * Creates a symlink ($link) for a $target. If the link folder does not exist it will be
+     * recursively craeted.
```
Comment (jdreesen) [inline tests/Installer/Command/InstallCommandTest.php:149]:
```suggestion
     * Mirrors the $src dir to the $dst dir (creates a backup of $dst first).
     *
```

File: tests/Installer/Command/InstallCommandTest.php [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+     * @param string $src
+     * @param string $dst
+     */
+    private function installMigrations(string $src, string $dst): void
+    {
+        $this->disposeFile($dst);
+        $this->filesystem->mkdir(dirname($dst));
```
Comment (jdreesen) [inline tests/Installer/Command/InstallCommandTest.php:156]: I'm not sure this is necessary, as `mirror()` internally calls [`$this->mkdir($targetDir)`](https://github.com/symfony/filesystem/blob/4.4/Filesystem.php#L564).

Comment (yariksheptykin) [reply tests/Installer/Command/InstallCommandTest.php:156]: good catch, @jdreesen !

**Comment by yariksheptykin** [comment]
Hi @dvesh3 ! Looks like this PR is stuck in the review pipeline. Anything I can do to help you with it? Feel free to reach out for explanations, more details, or changes.

Also, if you don't feel secure merging this in, feel free to close the PR. We can work out another way to get more test coverage.

**Comment by dvesh3** [comment]
Hello @yariksheptykin! It is planned for 6.8.0 milestone, so we'll review it soon 😉. thanks for extending help🙏

**Comment by yariksheptykin** [comment]
To me it does not seem that there is interest in this work. The branch is now stale and conflicts with the upstream changes. I am closing this MR to get this off your TODO-List, @dvesh3 . If you at some point decide to re-visit the topic of test-coverage feel free to reach out. I will be happy to contribute.

**Comment by dvesh3** [comment]
@yariksheptykin we needed more time to review this PR, so assigned it to 10.0.0 milestone(before: 7.0.0) and now we moved our test to Github Actions. I think we'll definitely re-visit this PR, when installer tests are in place 😉.

we really appreciate your time and efforts on this ❤️

**Pull request closed without merging**